builtin/game/falling.lua

```diff
@@ -146,29 +146,27 @@ end
 
 local function drop_attached_node(p)
 	local n = core.get_node(p)
-	local node_meta = core.get_meta(p)
-	local meta_fields = node_meta and node_meta:to_table().fields
+	local drops = core.get_node_drops(n, "")
+	local def = core.registered_items[n.name]
+	if def and def.preserve_metadata then
+		local oldmeta = core.get_meta(p):to_table().fields
+		-- Copy pos and node because the callback can modify them.
+		local pos_copy = {x=p.x, y=p.y, z=p.z}
+		local node_copy = {name=n.name, param1=n.param1, param2=n.param2}
+		local drop_stacks = {}
+		for k, v in pairs(drops) do
+			drop_stacks[k] = ItemStack(v)
+		end
+		drops = drop_stacks
+		def.preserve_metadata(pos_copy, node_copy, oldmeta, drops)
```

Comment:
problem i see with this callback: it can't work for falling nodes, because there is no `pos` at which the node actually is

Reply:
For an attached node, pos is the position that the node is in just before it becomes detached, which is when the callback occurs.  I actually think that a lot of nodes will not actually care about pos, they will just look at oldmeta.  But pos is still accurate.

For nodes that are members of the group "falling_node", as far as I am aware the problem does not exist - I believe that the metadata is intact when the node hits the ground, as the node is never converted into an item.

By the way, there is some [example code](https://gist.github.com/ashtrayoz/aad4a50d31b5dc8d0e0c0f9992d8e178) that you can play with if you want.

Reply:
But a falling node is an entity, so pos will not point to a node so that `core.get_node(pos).name == currently_falling_node`, which would be misleading to modders.
Am I missing something?

Reply:
When this is called, the node has not yet started to fall.  When I put a log-line into the example engraved_stone, I get:
[Server]: In preserve_metadata(), pos is now "(4,1,8)", node is now {
        param1 = 136,
        name = "test:engraved_stone",
        param2 = 0
}
So it seems okay from this end.

Reply:
Ah, node drops are calculated when the node starts to fall. That's fine then.